File: src/main/java/com/naminhyeok/fantazzk/room/application/CreateRoom.java
package com.naminhyeok.fantazzk.room.application;

import com.naminhyeok.fantazzk.CoreException;
import com.naminhyeok.fantazzk.room.domain.Room;
import com.naminhyeok.fantazzk.room.domain.RoomErrorType;
import com.naminhyeok.fantazzk.room.domain.RoomTeamLeader;
import com.naminhyeok.fantazzk.room.domain.RoomTemplateSpec;
import com.naminhyeok.fantazzk.room.domain.TeamLeaderId;
import com.naminhyeok.fantazzk.template.TemplateCatalog;
import java.time.Clock;
import java.time.Instant;
import java.util.UUID;
import java.util.regex.Pattern;
import lombok.RequiredArgsConstructor;
import org.hibernate.exception.ConstraintViolationException;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class CreateRoom {
    private static final int MAX_ROOM_CODE_ATTEMPTS = 3;
    private static final String ROOM_CODE_CONSTRAINT = "uk_rooms_code";
    private static final Pattern ROOM_CODE_CONSTRAINT_PATTERN = Pattern.compile(
        "(?i)(^|[^a-z0-9_])" + ROOM_CODE_CONSTRAINT + "([^a-z0-9_]|$)"
    );

    private final CreateRoomAttempt createRoomAttempt;
    private final TemplateCatalog templateCatalog;
    private final Clock clock;
    private final RoomCodeGenerator roomCodeGenerator;

    public RoomSessionResult create(UUID templateId, String hostNickname) {
        TemplateCatalog.TemplateBlueprint template = getTemplate(templateId);
        TeamLeaderId hostLeaderId = new TeamLeaderId(UUID.randomUUID().toString());
        String hostActionToken = UUID.randomUUID().toString();

        for (int attempt = 1; attempt <= MAX_ROOM_CODE_ATTEMPTS; attempt++) {
            Room room = newRoom(template, hostLeaderId, hostActionToken, hostNickname);
            RoomTeamLeader hostLeader = room.getLeaders().getFirst();
            try {
                Room saved = createRoomAttempt.save(room);
                return new RoomSessionResult(saved, hostLeader);
            } catch (DataIntegrityViolationException ex) {
                if (!isRoomCodeCollision(ex)) {
                    throw ex;
                }
                if (attempt == MAX_ROOM_CODE_ATTEMPTS) {
                    throw CoreException.of(RoomErrorType.ROOM_CODE_GENERATION_FAILED);
                }
            }
        }

        throw CoreException.of(RoomErrorType.ROOM_CODE_GENERATION_FAILED);
    }

    private TemplateCatalog.TemplateBlueprint getTemplate(UUID templateId) {
        try {
            return templateCatalog.getTemplate(templateId);
        } catch (TemplateCatalog.NotFound ex) {
            throw CoreException.of(RoomErrorType.ROOM_TEMPLATE_NOT_FOUND);
        }
    }

    private Room newRoom(
        TemplateCatalog.TemplateBlueprint template,
        TeamLeaderId hostLeaderId,
        String hostActionToken,
        String hostNickname
    ) {
        RoomTemplateSpec spec = RoomTemplateSpecFactory.from(template);
        return Room.createFromTemplate(
            generateCode(),
            hostLeaderId,
            hostNickname,
            hostActionToken,
            spec,
            Instant.now(clock)
        );
    }

    private String generateCode() {
        return roomCodeGenerator.generate();
    }

    private boolean isRoomCodeCollision(Throwable throwable) {
        Throwable current = throwable;
        while (current != null) {
            if (
                current instanceof ConstraintViolationException constraintViolationException &&
                isRoomCodeConstraint(constraintViolationException.getConstraintName())
            ) {
                return true;
            }
            current = current.getCause();
        }
        return false;
    }

    private boolean isRoomCodeConstraint(String constraintName) {
        return constraintName != null && ROOM_CODE_CONSTRAINT_PATTERN.matcher(constraintName).find();
    }
}

File: src/main/kotlin/com/naminhyeok/fantazzk/room/application/CreateRoom.kt
package com.naminhyeok.fantazzk.room.application

import com.naminhyeok.fantazzk.room.domain.Room
import com.naminhyeok.fantazzk.room.exception.RoomTemplateNotFoundException
import com.naminhyeok.fantazzk.room.repository.Rooms
import com.naminhyeok.fantazzk.template.TemplateBlueprint
import com.naminhyeok.fantazzk.template.TemplateCatalog
import com.naminhyeok.fantazzk.template.TemplateCatalogException
import com.naminhyeok.fantazzk.template.TemplateDraftOrderStrategy
import com.naminhyeok.fantazzk.template.TemplateId
import com.naminhyeok.fantazzk.template.TemplateMode
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.dao.DuplicateKeyException
import org.springframework.stereotype.Component
import java.util.UUID

@Component
class CreateRoom(
    private val roomRepository: Rooms,
    private val templateCatalog: TemplateCatalog,
    private val roomCreateAttemptExecutor: RoomCreateAttemptExecutor,
) {
    fun create(
        templateId: TemplateId,
        hostNickname: String,
    ): Room {
        val template =
            try {
                templateCatalog.get(templateId).toRoomTemplateSpec()
            } catch (_: TemplateCatalogException.NotFound) {
                throw RoomTemplateNotFoundException()
            } catch (_: TemplateCatalogException.Invalid) {
                throw IllegalStateException("유효하지 않은 템플릿입니다")
            }
        repeat(MAX_CODE_GENERATION_ATTEMPTS) {
            val code = generateCode()
            if (roomRepository.findByCode(code) != null) return@repeat

            val hostId = UUID.randomUUID().toString()
            val room =
                try {
                    roomCreateAttemptExecutor.create(
                        Room.createFromTemplate(
                            code = code,
                            hostId = hostId,
                            hostNickname = hostNickname,
                            spec = template,
                        ),
                    )
                } catch (_: DuplicateKeyException) {
                    return@repeat
                } catch (exception: DataIntegrityViolationException) {
                    if (roomRepository.findByCode(code) != null) {
                        return@repeat
                    }
                    throw exception
                }
            return room
        }
        throw IllegalStateException("방 코드를 생성할 수 없습니다")
    }

    private fun generateCode(): String {
        val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        return (1..6).map { chars.random() }.joinToString("")
    }

    companion object {
        private const val MAX_CODE_GENERATION_ATTEMPTS = 5
    }
}

private fun TemplateBlueprint.toRoomTemplateSpec(): RoomTemplateSpec =
    RoomTemplateSpec(
        mode =
            when (mode) {
                TemplateMode.AUCTION -> RoomTemplateSpec.Mode.AUCTION
                TemplateMode.DRAFT -> RoomTemplateSpec.Mode.DRAFT
            },
        teamCount = teamCount,
        teamSize = teamSize,
        budget = budget,
        draftOrderStrategy =
            when (draftOrderStrategy) {
                null -> null
                TemplateDraftOrderStrategy.FIXED -> RoomTemplateSpec.DraftOrderStrategy.FIXED
                TemplateDraftOrderStrategy.SNAKE -> RoomTemplateSpec.DraftOrderStrategy.SNAKE
            },
        players =
            players.map {
                RoomTemplateSpec.Player(
                    name = it.name,
                    displayOrder = it.displayOrder,
                )
            },
    )
